std/array.d

```diff
@@ -778,39 +778,64 @@ private auto arrayAllocImpl(bool minimallyInitialized, T, I...)(I sizes) nothrow
 }
 
 // overlap
-/*
-NOTE: Undocumented for now, overlap does not yet work with ctfe.
+
+
+/++
 Returns the overlapping portion, if any, of two arrays. Unlike $(D
-equal), $(D overlap) only compares the pointers in the ranges, not the
-values referred by them. If $(D r1) and $(D r2) have an overlapping
-slice, returns that slice. Otherwise, returns the null slice.
-*/
-auto overlap(T, U)(T[] r1, U[] r2) @trusted pure nothrow
-if (is(typeof(r1.ptr < r2.ptr) == bool))
+equal), $(D overlap) only compares the pointers and lengths in the
+ranges, not the values referred by them. If $(D r1) and $(D r2) have an
+overlapping slice, returns that slice. Otherwise, returns the null
+slice.
++/
+
+auto overlap(T, U)(T[] a, U[] b) @trusted
+if (is(typeof(a.ptr < b.ptr) == bool))
 {
-    import std.algorithm.comparison : min, max;
-    auto b = max(r1.ptr, r2.ptr);
-    auto e = min(r1.ptr + r1.length, r2.ptr + r2.length);
-    return b < e ? b[0 .. e - b] : null;
+    import std.algorithm.comparison : min;
+
+    alias RetPtr = CommonType!(T*, U*);
+
+    if (a.ptr <= b.ptr && b.ptr < a.ptr + a.length)
```

Comment:
@ZombineDev comparing is @safe, but the unsafe part here are the casts and pointer arithmetic.

@JackStouffer Bounds checking is intended for dereferencing, and we don't do that here. Now when I think it, it would be a performance slowdown. If the arrays point to invalid memory, safety has already failed somewhere, not here. It's typical user code that should bounds-check, but not low-level library functions like this, unless dereferencing.

What comes to the "compromise", that would not remove any surface for bugs to manifest, since the pointer arithmetic would still be there. It would only make the potential bug harder to find.